src/ctap2/commands/make_credentials.rs

```diff
@@ -0,0 +1,860 @@
+        map.serialize_entry(&1, &client_data_hash)?;
+        map.serialize_entry(&2, &self.rp)?;
+        map.serialize_entry(&3, &self.user)?;
+        map.serialize_entry(&4, &self.pub_cred_params)?;
+        if !self.exclude_list.is_empty() {
+            map.serialize_entry(&5, &self.exclude_list)?;
+        }
+        if !self.extensions.is_empty() {
+            map.serialize_entry(&6, &self.extensions)?;
+        }
+        if self.options.has_some() {
+            map.serialize_entry(&7, &self.options)?;
+        }
+        if let Some(pin_auth) = &self.pin_auth {
+            map.serialize_entry(&8, &pin_auth)?;
+            map.serialize_entry(&9, &1)?; // version
+        }
+        map.end()
+    }
+}
+
+impl Request<(AttestationObject, CollectedClientData)> for MakeCredentials {}
+
+impl RequestCtap1 for MakeCredentials {
+    type Output = (AttestationObject, CollectedClientData);
+
+    fn apdu_format<Dev>(&self, _dev: &mut Dev) -> Result<Vec<u8>, HIDError>
+    where
+        Dev: U2FDevice,
+    {
+        // TODO(MS): Mandatory sanity checks are missing:
+        // https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#u2f-authenticatorMakeCredential-interoperability
+        // If any of the below conditions is not true, platform errors out with CTAP2_ERR_UNSUPPORTED_OPTION.
+        //  * pubKeyCredParams must use the ES256 algorithm (-7).
+        //  * Options must not include "rk" set to true.
+        //  * Options must not include "uv" set to true.
+
+        let flags = if self.options.ask_user_validation() {
+            U2F_REQUEST_USER_PRESENCE
+        } else {
+            0
+        };
+
+        let mut register_data = Vec::with_capacity(2 * PARAMETER_SIZE);
+        register_data.extend_from_slice(
+            self.client_data
+                .hash()
+                .map_err(|e| HIDError::Command(CommandError::Json(e)))?
+                .as_ref(),
+        );
+        register_data.extend_from_slice(self.rp.hash().as_ref());
+        let cmd = U2F_REGISTER;
+        let apdu = U2FAPDUHeader::serialize(cmd, flags, &register_data)?;
+
+        Ok(apdu)
+    }
+
+    fn handle_response_ctap1(
+        &self,
+        status: Result<(), ApduErrorStatus>,
+        input: &[u8],
+    ) -> Result<Self::Output, Retryable<HIDError>> {
+        if Err(ApduErrorStatus::ConditionsNotSatisfied) == status {
+            return Err(Retryable::Retry);
+        }
+
+        named!(
+            parse_register<(&[u8], &[u8])>,
+            do_parse!(
+                reserved: tag!(&[0x05])
+                    >> public_key: take!(65)
+                    >> key_handle_len: be_u8
+                    >> key_handle: take!(key_handle_len)
+                    >> (public_key, key_handle)
+            )
+        );
+
+        let (rest, (public_key, key_handle)) = parse_register(input)
+            .map_err(|e| {
```

Comment:
These `map_err` calls are redundant, right? Can we remove them and just return Retryable::Error in the first one?